File: python/l402kit/replay.py
import hashlib
from typing import Any

_used_preimages: set[str] = set()


def check_and_mark_preimage(preimage: str) -> bool:
    """In-memory replay protection (single process). Resets on restart."""
    key = hashlib.sha256(preimage.encode()).hexdigest()
    if key in _used_preimages:
        return False
    _used_preimages.add(key)
    return True


class MemoryReplayAdapter:
    """In-memory replay store. Zero dependencies, single-process only."""

    def __init__(self) -> None:
        self._used: set[str] = set()

    def check_and_mark(self, preimage: str) -> bool:
        key = hashlib.sha256(preimage.encode()).hexdigest()
        if key in self._used:
            return False
        self._used.add(key)
        return True


class RedisReplayAdapter:
    """Redis-backed replay protection for multi-instance / production deployments.

    Usage:
        import redis
        from l402kit.replay import RedisReplayAdapter

        r = redis.Redis.from_url(os.environ["REDIS_URL"])
        adapter = RedisReplayAdapter(r)

        # Pass to middleware:
        l402_middleware(price_sats=10, lightning=provider, replay=adapter)
    """

    def __init__(self, redis_client: Any, ttl_seconds: int = 86400) -> None:
        self._redis = redis_client
        self._ttl = ttl_seconds

    def check_and_mark(self, preimage: str) -> bool:
        """Returns True on first use, False if already seen (replay attack)."""
        key = "l402:used:" + hashlib.sha256(preimage.encode()).hexdigest()
        # SET key 1 NX EX ttl — atomic: only sets if key doesn't exist
        result = self._redis.set(key, "1", nx=True, ex=self._ttl)
        return result is not None

File: src/replay.ts
import { createHash } from "crypto";

// ─── Interface ────────────────────────────────────────────────────────────────

/**
 * Pluggable replay-protection adapter.
 * Implement this interface to use any storage backend.
 */
export interface ReplayAdapter {
  /**
   * Returns `true` if this preimage was NOT yet seen and marks it as used.
   * Returns `false` if it was already seen (replay attack).
   * Must be atomic — concurrent calls with the same preimage must not both return `true`.
   */
  check(preimage: string): Promise<boolean> | boolean;
}

// ─── Memory adapter (default) ─────────────────────────────────────────────────

/**
 * In-process replay protection. Fast, zero dependencies.
 * **Not** safe across multiple server instances or process restarts.
 * Use `RedisReplayAdapter` for production multi-instance deployments.
 */
export class MemoryReplayAdapter implements ReplayAdapter {
  private readonly seen = new Set<string>();

  check(preimage: string): boolean {
    const key = createHash("sha256").update(preimage).digest("hex");
    if (this.seen.has(key)) return false;
    this.seen.add(key);
    return true;
  }

  /** Clear all state — useful in tests. */
  clear(): void { this.seen.clear(); }
}

// ─── Redis adapter ────────────────────────────────────────────────────────────

/**
 * Minimal Redis interface — compatible with `ioredis` and `@upstash/redis`.
 * If you use `node-redis` v4, wrap it:
 * ```ts
 * { set: (k, v, ex, ttl, nx) => client.set(k, v, { EX: ttl, NX: true }) }
 * ```
 */
export interface RedisLike {
  set(
    key: string,
    value: string,
    exMode: "EX",
    ttlSecs: number,
    nxMode: "NX",
  ): Promise<"OK" | null | string | undefined>;
}

/**
 * Distributed replay protection backed by Redis.
 * Uses atomic `SET NX` — safe across multiple server instances and restarts.
 *
 * Compatible with: `ioredis`, `@upstash/redis`, `node-redis` (with wrapper)
 *
 * @example
 * ```ts
 * import Redis from "ioredis";
 * import { l402, RedisReplayAdapter } from "l402-kit";
 *
 * const redis = new Redis(process.env.REDIS_URL);
 *
 * app.use("/premium", l402({
 *   priceSats: 100,
 *   replayAdapter: new RedisReplayAdapter(redis),
 * }));
 * ```
 */
export class RedisReplayAdapter implements ReplayAdapter {
  constructor(
    private readonly redis: RedisLike,
    /** How long to remember a preimage. Default: 7 days (covers invoice expiry + buffer). */
    private readonly ttlSecs: number = 86_400 * 7,
  ) {}

  async check(preimage: string): Promise<boolean> {
    const key = `l402:replay:${createHash("sha256").update(preimage).digest("hex")}`;
    // Atomic SET NX: sets the key only if it does not exist.
    // Returns "OK" on first write (not seen), null if key already existed (replay).
    const result = await this.redis.set(key, "1", "EX", this.ttlSecs, "NX");
    return result !== null && result !== undefined;
  }
}

// ─── Module-level default (backwards compat) ─────────────────────────────────

const _default = new MemoryReplayAdapter();

/**
 * Check and mark a preimage using the default in-memory store.
 * @deprecated Pass a `ReplayAdapter` via `l402({ replayAdapter })` instead.
 * This module-level store does not survive restarts or scale across instances.
 */
export function checkAndMarkPreimage(preimage: string): boolean {
  return _default.check(preimage);
}
